openssl/ecdh.go

```diff
@@ -69,144 +84,123 @@ func NewPrivateKeyECDH(curve string, bytes []byte) (*PrivateKeyECDH, error) {
 	if err != nil {
 		return nil, err
 	}
+	b := bytesToBN(bytes)
+	if b == nil {
+		return nil, newOpenSSLError("BN_bin2bn failed")
+	}
+	defer C.go_openssl_BN_free(b)
 	key := C.go_openssl_EC_KEY_new_by_curve_name(nid)
 	if key == nil {
 		return nil, newOpenSSLError("EC_KEY_new_by_curve_name")
 	}
-	b := bytesToBN(bytes)
-	ok := b != nil && C.go_openssl_EC_KEY_set_private_key(key, b) != 0
-	if b != nil {
-		C.go_openssl_BN_free(b)
-	}
-	if !ok {
+	if C.go_openssl_EC_KEY_set_private_key(key, b) != 1 {
 		C.go_openssl_EC_KEY_free(key)
 		return nil, newOpenSSLError("EC_KEY_set_private_key")
 	}
-	k := &PrivateKeyECDH{curve, key}
+	pkey, err := newEVPPKEY(key)
+	if err != nil {
+		C.go_openssl_EC_KEY_free(key)
+		return nil, err
+	}
+	k := &PrivateKeyECDH{pkey}
 	runtime.SetFinalizer(k, (*PrivateKeyECDH).finalize)
 	return k, nil
 }
 
 func (k *PrivateKeyECDH) PublicKey() (*PublicKeyECDH, error) {
 	defer runtime.KeepAlive(k)
-
-	group := C.go_openssl_EC_KEY_get0_group(k.key)
+	key := C.go_openssl_EVP_PKEY_get1_EC_KEY(k._pkey)
+	if key == nil {
+		return nil, newOpenSSLError("EVP_PKEY_get1_EC_KEY")
+	}
+	defer C.go_openssl_EC_KEY_free(key)
+	group := C.go_openssl_EC_KEY_get0_group(key)
 	if group == nil {
 		return nil, newOpenSSLError("EC_KEY_get0_group")
 	}
-	kbig := C.go_openssl_EC_KEY_get0_private_key(k.key)
-	if kbig == nil {
-		return nil, newOpenSSLError("EC_KEY_get0_private_key")
-	}
-	pt := C.go_openssl_EC_POINT_new(group)
+	pt := C.go_openssl_EC_KEY_get0_public_key(key)
 	if pt == nil {
-		return nil, newOpenSSLError("EC_POINT_new")
-	}
-	if C.go_openssl_EC_POINT_mul(group, pt, kbig, nil, nil, nil) == 0 {
-		C.go_openssl_EC_POINT_free(pt)
-		return nil, newOpenSSLError("EC_POINT_mul")
+		// The public key will be nil if k has been generated using
+		// NewPrivateKeyECDH instead of GenerateKeyECDH.
+		// OpenSSL does not expose any method to generate the public
+		// key from the private key, so we have to calculate it here.
+		// Notice that this path bypasses OpenSSL FIPS module.
```

Comment:
I can't really tell what crosses the line. It seems roughly reasonable that doing this calculation could be against FIPS in some way, but I don't see the direct link. (This is using OpenSSL-provided `EC_POINT_mul` after all...)

Reply:
Agree. I've removed the notice line and added a link to the OpenSSL issue where a maintainer states that there is no way to generate the public key from a private one using the public OpenSSL api.